experimental/oak_async/src/executor.rs

```diff
@@ -0,0 +1,176 @@
+use futures::{executor::LocalPool, task::LocalSpawnExt};
+use log::{debug, trace};
+use oak::{ChannelReadStatus, Handle, OakStatus, ReadHandle};
+use std::collections::HashMap;
+
+// thread local so `cargo test` can run tests in parallel.
+// Oak nodes are single-threaded.
+std::thread_local! {
+    static EXECUTOR: RefCell<Executor> = RefCell::new(Executor::default());
+}
+
+type ReaderId = usize;
+
+struct WaitingReader {
+    handle: Handle,
+    waker: Waker,
+}
+
+/// Global executor context.
+///
+/// `with_executor` provides a way to get a handle to the global executor instance.
+#[derive(Default)]
+pub struct Executor {
+    /// Used to assign a unique id to each reader. It is incremented by every newly created reader
+    reader_id_counter: ReaderId,
+    /// Set of readers waiting for data. All handles in this map are polled in `wait_on_channels`.
+    // Why not HashMap<Handle, Vec<Waker>>? When a future asks to be removed from the waiting list
+    // (i.e. when it is dropped), we need to be able to remove their Waker from the map.
+    waiting_readers: HashMap<ReaderId, WaitingReader>,
+}
+
+impl Executor {
+    pub fn add_waiting_reader(&mut self, reader_id: usize, handle: ReadHandle, waker: &Waker) {
+        trace!(
+            "Add waiting reader {} waiting on handle {}",
+            reader_id,
+            handle.handle
+        );
+        let _ = self.waiting_readers.insert(
+            reader_id,
+            WaitingReader {
+                handle: handle.handle,
+                waker: waker.clone(),
+            },
+        );
+    }
+
+    pub fn remove_waiting_reader(&mut self, id: usize) {
+        trace!("Remove waiting reader {}", id);
+        if self.waiting_readers.remove(&id).is_none() {
+            // This is usually not an error. If a Future is dropped as a result of it being woken up
+            // and then resolving, we expect the reader_id to not be present in the waiting set.
+            debug!(
+                "Attempted to remove waiting reader {}, but no such reader was in the waiting set",
+                id
+            )
+        }
+    }
+
+    pub fn new_id(&mut self) -> usize {
+        let id = self.reader_id_counter;
+        // Collissions are technically possible, but only if one read is very slow and many async
+        // reads are requested in quick succession.
+        self.reader_id_counter = self.reader_id_counter.wrapping_add(1);
+        id
+    }
+
+    pub fn none_waiting(&self) -> bool {
+        self.waiting_readers.is_empty()
+    }
+
+    pub fn pending_handles(&self) -> HashMap<Handle, ReaderId> {
+        self.waiting_readers
+            .iter()
+            .map(|(reader_id, waiting_reader)| (waiting_reader.handle, *reader_id))
+            .collect()
+    }
+
+    /// Remove a reader from the waiting set and wake it
+    pub fn wake_reader(&mut self, reader_id: usize) {
+        self.waiting_readers
+            .remove(&reader_id)
+            .unwrap()
```

Comment:
Can this not happen concurrently, in which case a remove call may return None?

Reply:
I'm not sure what you mean with concurrently? The executor loop is what calls this function, and when we're in that loop no `ChannelRead` can become dropped and remove itself.

Reply:
Makes sense, and perhaps what you just replied may be part of a message to `expect`, just in case it does happen because of some future bug in the executor loop. Up to you anyways.